File: cricksense/pipeline/players_meta.py
"""
Pakistan player metadata: role classification, current-squad flag, captaincy,
and display-name overrides.

Cricsheet gives us names and ball-by-ball appearances only -- no role, squad
status, or captaincy. This table is manually curated from public cricketing
knowledge for every Pakistan player who appears in a Test from 2021 onwards.
Flagged for human review in NEEDS_YOUR_INPUT.md: squad status and captaincy
drift over time (retirements, selection changes) and should be reviewed
periodically -- this snapshot reflects the squad as of mid-2026.

role must be one of: batsman, wicketkeeper, fast_bowler, spinner, all_rounder
(matches the player_role enum in src/db/schema.ts).
"""

# name -> (role, role_label, is_current_squad, is_captain, display_name)
PLAYERS_META: dict[str, tuple[str, str, bool, bool, str]] = {
    "Aamer Jamal":          ("all_rounder",  "All-rounder",  False, False, "Aamer Jamal"),
    "Abdullah Fazal":       ("batsman",      "Batsman",      False, False, "Abdullah Fazal"),
    "Abdullah Shafique":    ("batsman",      "Opener",       True,  False, "Abdullah Shafique"),
    "Abid Ali":             ("batsman",      "Opener",       False, False, "Abid Ali"),
    "Abrar Ahmed":          ("spinner",      "Spinner",      True,  False, "Abrar Ahmed"),
    "Agha Salman":          ("all_rounder",  "All-rounder",  True,  False, "Salman Agha"),
    "Ali Usman":            ("batsman",      "Batsman",      False, False, "Ali Usman"),
    "Asif Afridi":          ("spinner",      "Spinner",      False, False, "Asif Afridi"),
    "Awais Zafar":          ("batsman",      "Batsman",      False, False, "Awais Zafar"),
    "Azan Awais":           ("batsman",      "Opener",       False, False, "Azan Awais"),
    "Azhar Ali":            ("batsman",      "Batsman",      False, False, "Azhar Ali"),
    "Babar Azam":           ("batsman",      "Batsman",      True,  False, "Babar Azam"),
    "Faheem Ashraf":        ("all_rounder",  "All-rounder",  False, False, "Faheem Ashraf"),
    "Fawad Alam":           ("batsman",      "Batsman",      False, False, "Fawad Alam"),
    "Haris Rauf":           ("fast_bowler",  "Fast bowler",  False, False, "Haris Rauf"),
    "Haris Sohail":         ("batsman",      "Batsman",      False, False, "Haris Sohail"),
    "Hasan Ali":            ("fast_bowler",  "Fast bowler",  False, False, "Hasan Ali"),
    "Iftikhar Ahmed":       ("all_rounder",  "All-rounder",  False, False, "Iftikhar Ahmed"),
    "Imam-ul-Haq":          ("batsman",      "Opener",       False, False, "Imam-ul-Haq"),
    "Imran Butt":           ("batsman",      "Opener",       False, False, "Imran Butt"),
    "Kamran Ghulam":        ("batsman",      "Batsman",      False, False, "Kamran Ghulam"),
    "Kashif Ali":           ("fast_bowler",  "Fast bowler",  False, False, "Kashif Ali"),
    "Khurram Shahzad":      ("fast_bowler",  "Fast bowler",  False, False, "Khurram Shahzad"),
    "Mir Hamza":            ("fast_bowler",  "Fast bowler",  True,  False, "Mir Hamza"),
    "Mohammad Abbas":       ("fast_bowler",  "Fast bowler",  False, False, "Mohammad Abbas"),
    "Mohammad Ali":         ("fast_bowler",  "Fast bowler",  False, False, "Mohammad Ali"),
    "Mohammad Huraira":     ("batsman",      "Batsman",      False, False, "Mohammad Huraira"),
    "Mohammad Nawaz":       ("all_rounder",  "All-rounder",  False, False, "Mohammad Nawaz"),
    "Mohammad Rizwan":      ("wicketkeeper", "Wicketkeeper", True,  False, "Mohammad Rizwan"),
    "Mohammad Wasim":       ("fast_bowler",  "Fast bowler",  False, False, "Mohammad Wasim"),
    "Naseem Shah":          ("fast_bowler",  "Fast bowler",  True,  False, "Naseem Shah"),
    "Nauman Ali":           ("spinner",      "Spinner",      True,  False, "Noman Ali"),
    "Saim Ayub":            ("batsman",      "Opener",       False, False, "Saim Ayub"),
    "Sajid Khan":           ("spinner",      "Spinner",      True,  False, "Sajid Khan"),
    "Sarfraz Ahmed":        ("wicketkeeper", "Wicketkeeper", False, False, "Sarfraz Ahmed"),
    "Saud Shakeel":         ("batsman",      "Batsman",      True,  False, "Saud Shakeel"),
    "Shaheen Shah Afridi":  ("fast_bowler",  "Fast bowler",  True,  False, "Shaheen Afridi"),
    "Shan Masood":          ("batsman",      "Opener · Captain", True, True, "Shan Masood"),
    "Tabish Khan":          ("fast_bowler",  "Fast bowler",  False, False, "Tabish Khan"),
    "Ubaid Shah":           ("fast_bowler",  "Fast bowler",  False, False, "Ubaid Shah"),
    "Yasir Shah":           ("spinner",      "Spinner",      False, False, "Yasir Shah"),
    "Zafar Gohar":          ("spinner",      "Spinner",      False, False, "Zafar Gohar"),
    "Zahid Mahmood":        ("spinner",      "Spinner",      False, False, "Zahid Mahmood"),
}

DEFAULT_META = ("batsman", "Batsman", False, False)


def get_player_meta(cricsheet_name: str) -> tuple[str, str, bool, bool, str]:
    if cricsheet_name in PLAYERS_META:
        return PLAYERS_META[cricsheet_name]
    role, label, current, captain = DEFAULT_META
    return (role, label, current, captain, cricsheet_name)

File: cricksense/src/lib/pipeline/players-meta.ts
// Ported from pipeline/players_meta.py -- keep both in sync if you edit either.
// See that file's docstring for the full rationale (Cricsheet has no role,
// squad-status, or captaincy field; this is a manually curated snapshot).
// Unrecognised names default to a plain batsman, not in the current squad --
// same fallback as the Python version -- so a debutant shows up immediately
// with sane defaults rather than breaking the pipeline, pending a manual fix.

export type PlayerRole = "batsman" | "wicketkeeper" | "fast_bowler" | "spinner" | "all_rounder";

export type PlayerMeta = {
  role: PlayerRole;
  roleLabel: string;
  isCurrentSquad: boolean;
  isCaptain: boolean;
  displayName: string;
};

const PLAYERS_META: Record<string, [PlayerRole, string, boolean, boolean, string]> = {
  "Aamer Jamal": ["all_rounder", "All-rounder", false, false, "Aamer Jamal"],
  "Abdullah Fazal": ["batsman", "Batsman", false, false, "Abdullah Fazal"],
  "Abdullah Shafique": ["batsman", "Opener", true, false, "Abdullah Shafique"],
  "Abid Ali": ["batsman", "Opener", false, false, "Abid Ali"],
  "Abrar Ahmed": ["spinner", "Spinner", true, false, "Abrar Ahmed"],
  "Agha Salman": ["all_rounder", "All-rounder", true, false, "Salman Agha"],
  "Ali Usman": ["batsman", "Batsman", false, false, "Ali Usman"],
  "Asif Afridi": ["spinner", "Spinner", false, false, "Asif Afridi"],
  "Awais Zafar": ["batsman", "Batsman", false, false, "Awais Zafar"],
  "Azan Awais": ["batsman", "Opener", false, false, "Azan Awais"],
  "Azhar Ali": ["batsman", "Batsman", false, false, "Azhar Ali"],
  "Babar Azam": ["batsman", "Batsman", true, false, "Babar Azam"],
  "Faheem Ashraf": ["all_rounder", "All-rounder", false, false, "Faheem Ashraf"],
  "Fawad Alam": ["batsman", "Batsman", false, false, "Fawad Alam"],
  "Haris Rauf": ["fast_bowler", "Fast bowler", false, false, "Haris Rauf"],
  "Haris Sohail": ["batsman", "Batsman", false, false, "Haris Sohail"],
  "Hasan Ali": ["fast_bowler", "Fast bowler", false, false, "Hasan Ali"],
  "Iftikhar Ahmed": ["all_rounder", "All-rounder", false, false, "Iftikhar Ahmed"],
  "Imam-ul-Haq": ["batsman", "Opener", false, false, "Imam-ul-Haq"],
  "Imran Butt": ["batsman", "Opener", false, false, "Imran Butt"],
  "Kamran Ghulam": ["batsman", "Batsman", false, false, "Kamran Ghulam"],
  "Kashif Ali": ["fast_bowler", "Fast bowler", false, false, "Kashif Ali"],
  "Khurram Shahzad": ["fast_bowler", "Fast bowler", false, false, "Khurram Shahzad"],
  "Mir Hamza": ["fast_bowler", "Fast bowler", true, false, "Mir Hamza"],
  "Mohammad Abbas": ["fast_bowler", "Fast bowler", false, false, "Mohammad Abbas"],
  "Mohammad Ali": ["fast_bowler", "Fast bowler", false, false, "Mohammad Ali"],
  "Mohammad Huraira": ["batsman", "Batsman", false, false, "Mohammad Huraira"],
  "Mohammad Nawaz": ["all_rounder", "All-rounder", false, false, "Mohammad Nawaz"],
  "Mohammad Rizwan": ["wicketkeeper", "Wicketkeeper", true, false, "Mohammad Rizwan"],
  "Mohammad Wasim": ["fast_bowler", "Fast bowler", false, false, "Mohammad Wasim"],
  "Naseem Shah": ["fast_bowler", "Fast bowler", true, false, "Naseem Shah"],
  "Nauman Ali": ["spinner", "Spinner", true, false, "Noman Ali"],
  "Saim Ayub": ["batsman", "Opener", false, false, "Saim Ayub"],
  "Sajid Khan": ["spinner", "Spinner", true, false, "Sajid Khan"],
  "Sarfraz Ahmed": ["wicketkeeper", "Wicketkeeper", false, false, "Sarfraz Ahmed"],
  "Saud Shakeel": ["batsman", "Batsman", true, false, "Saud Shakeel"],
  "Shaheen Shah Afridi": ["fast_bowler", "Fast bowler", true, false, "Shaheen Afridi"],
  "Shan Masood": ["batsman", "Opener · Captain", true, true, "Shan Masood"],
  "Tabish Khan": ["fast_bowler", "Fast bowler", false, false, "Tabish Khan"],
  "Ubaid Shah": ["fast_bowler", "Fast bowler", false, false, "Ubaid Shah"],
  "Yasir Shah": ["spinner", "Spinner", false, false, "Yasir Shah"],
  "Zafar Gohar": ["spinner", "Spinner", false, false, "Zafar Gohar"],
  "Zahid Mahmood": ["spinner", "Spinner", false, false, "Zahid Mahmood"],
};

export function getPlayerMeta(cricsheetName: string): PlayerMeta {
  const entry = PLAYERS_META[cricsheetName];
  if (entry) {
    const [role, roleLabel, isCurrentSquad, isCaptain, displayName] = entry;
    return { role, roleLabel, isCurrentSquad, isCaptain, displayName };
  }
  return { role: "batsman", roleLabel: "Batsman", isCurrentSquad: false, isCaptain: false, displayName: cricsheetName };
}
